### PR TITLE
Add support for Solidity

### PR DESCRIPTION
**Pull request by shekhirin**
Resolves https://github.com/compiler-explorer/compiler-explorer/issues/2723

| Compilation | Screenshot
| - | -
| Successful | image
| Failed | image

### CODE REVIEW

**Comment by mattgodbolt** [comment]
Nice! Have you seen our [docs on how to add compilers](https://github.com/compiler-explorer/compiler-explorer/blob/main/docs/AddingACompiler.md#adding-a-new-compiler-to-the-live-site) - it's a bit out of date as now our compilers are installed using yaml in https://github.com/compiler-explorer/infra/tree/main/bin/yaml - hopefully installing the solidity compiler is straightforward!

**Comment by shekhirin** [comment]
@mattgodbolt yeah, I've seen it but I'm afraid no installable type satisfies what I need for Solidity: they distribute static linux binaries without tarballs/compression (as far as I've seen on https://docs.soliditylang.org/en/latest/installing-solidity.html#static-binaries).

So I guess we can add `BinaryInstallable` type?

**Comment by mattgodbolt** [comment]
Sounds like a plan @shekhirin - I can look at adding that. Should be pretty quick to do :) - I'll take a look in a bit!

**Comment by mattgodbolt** [comment]
I can only find binaries going back to 0.4.10 on https://binaries.soliditylang.org/linux-amd64/list.json -- is there another location for installations?

**Comment by mattgodbolt** [comment]
https://github.com/compiler-explorer/infra/pull/705 installs everything available from the list.json; and should be able to be modified to install other versions by URL too with `singleFile` type.

File: etc/config/solidity.defaults.properties [hunk 1]
```diff
@@ -0,0 +1,11 @@
+compilerType=solidity
+supportsBinary=false
+instructionSet=evm
+
+group.solc.isSemVer=true
+group.solc.baseName=solc
```
Comment (RubenRBS) [inline etc/config/solidity.defaults.properties:6]: Can you add a default `/usr/bin/solc` entry for the default env? That's how it's usually done for the rest of the languages

File: etc/config/solidity.amazon.properties [hunk 2]
```diff
@@ -0,0 +1,175 @@
+compilers=&solc
+defaultCompiler=solc0813
+
+group.solc.compilers=solc036:solc040:solc041:solc042:solc043:solc044:solc045:solc046:solc047:solc048:solc049:solc0410:solc0411:solc0412:solc0413:solc0414:solc0415:solc0416:solc0417:solc0418:solc0419:solc0420:solc0421:solc0422:solc0423:solc0424:solc0425:solc0426:solc050:solc051:solc052:solc053:solc054:solc055:solc056:solc057:solc058:solc059:solc0510:solc0511:solc0512:solc0513:solc0514:solc0515:solc0516:solc0517:solc060:solc061:solc062:solc063:solc064:solc065:solc066:solc067:solc068:solc069:solc0610:solc0611:solc0612:solc070:solc071:solc072:solc073:solc074:solc075:solc076:solc080:solc081:solc082:solc083:solc084:solc085:solc086:solc087:solc088:solc089:solc0810:solc0811:solc0812:solc0813
```
Comment (RubenRBS) [inline etc/config/solidity.amazon.properties:4]: Can you add all 3
```
compilerType=solidity
supportsBinary=false
instructionSet=evm
```
for this group in this file?

**Review by RubenRBS — CHANGES_REQUESTED** [review]
Looking good

File: etc/config/solidity.defaults.properties (hunk 1 above)
Comment (shekhirin) [reply etc/config/solidity.defaults.properties:7]: done

File: etc/config/solidity.amazon.properties (hunk 2 above)
Comment (shekhirin) [reply etc/config/solidity.amazon.properties:4]: done

**Comment by shekhirin** [comment]
> I can only find binaries going back to 0.4.10 on https://binaries.soliditylang.org/linux-amd64/list.json -- is there another location for installations?

I think we can go for these versions only for now, afaik earlier ones are more tricky to discover.

File: etc/config/solidity.defaults.properties [hunk 3]
```diff
@@ -0,0 +1,12 @@
+compilers=/usr/bin/solc
+compilerType=solidity
+supportsBinary=false
+instructionSet=evm
+
+group.solc.isSemVer=true
+group.solc.baseName=solc
```
Comment (RubenRBS) [inline etc/config/solidity.defaults.properties:7]:
```suggestion
isSemVer=true
baseName=solc
```

**Review by RubenRBS — APPROVED** [review]

**Comment by RubenRBS** [comment]
Thanks again! I'll merge now, fix some extra issues, and see if we can push and update to the site Soon ™️

**Comment by anders-torbjornsen** [comment]
Great work @shekhirin, I literally just forked compiler explorer with the intention to have a go at doing this, and saw the commit message in the repo!

**Comment by mattgodbolt** [comment]
We're missing solidity 0.3.6 so I've removed that from the installation scripts (in the infra repo). Anyone here know where we might find it though?

**Comment by mattgodbolt** [comment]
```
Apr/14 13:33 admin-node~/infra (main|✔) 1.2s $ sudo ce_install install solidity                                                             
2022-04-14 13:33:53,279 lib.installation INFO     Making uncached requests                                                                  
Installing compilers/solidity 0.4.26                                                                                                        
2022-04-14 13:33:56,371 lib.installation INFO     Fetching https://binaries.soliditylang.org/linux-amd64/solc-linux-amd64-v0.4.26+commit.456
3c3fc (5235472 bytes)                                                                                                                       
2022-04-14 13:33:56,400 lib.installation INFO     100% of https://binaries.soliditylang.org/linux-amd64/solc-linux-amd64-v0.4.26+commit.4563
c3fc                                                                                                                                        
2022-04-14 13:33:56,507 lib.installation INFO     Moving from staging (/opt/compiler-explorer/staging/solc-0.4.26) to final destination (/op
t/compiler-explorer/solc-0.4.26)                                                                                                            
2022-04-14 13:33:56,655 lib.installation INFO     compilers/solidity 0.4.26 installed OK                                                    
Installing compilers/solidity 0.5.17                                                                                                        
2022-04-14 13:33:56,700 lib.installation INFO     Fetching https://binaries.soliditylang.org/linux-amd64/solc-linux-amd64-v0.5.17+commit.d19
bba13 (8007728 bytes)                                                                                                                       
2022-04-14 13:33:56,739 lib.installation INFO     100% of https://binaries.soliditylang.org/linux-amd64/solc-linux-amd64-v0.5.17+commit.d19b
ba13                                                                                                                                        
2022-04-14 13:33:56,864 lib.installation INFO     Moving from staging (/opt/compiler-explorer/staging/solc-0.5.17) to final destination (/op
t/compiler-explorer/solc-0.5.17)                                                                                                            
2022-04-14 13:33:57,048 lib.installation INFO     compilers/solidity 0.5.17 installed OK                                                    
Installing compilers/solidity 0.6.12                                                                                                        
2022-04-14 13:33:57,098 lib.installation INFO     Fetching https://binaries.soliditylang.org/linux-amd64/solc-linux-amd64-v0.6.12+commit.27d
51765 (9474128 bytes)                                                                                                                       
2022-04-14 13:33:57,136 lib.installation INFO     100% of https://binaries.soliditylang.org/linux-amd64/solc-linux-amd64-v0.6.12+commit.27d5
1765                                                                                                                                        
2022-04-14 13:33:57,309 lib.installation INFO     Moving from staging (/opt/compiler-explorer/staging/solc-0.6.12) to final destination (/op
t/compiler-explorer/solc-0.6.12)                                                                                                            
2022-04-14 13:33:57,494 lib.installation INFO     compilers/solidity 0.6.12 installed OK                                                    
Installing compilers/solidity 0.7.6                                                                                                         
2022-04-14 13:33:57,541 lib.installation INFO     Fetching https://binaries.soliditylang.org/linux-amd64/solc-linux-amd64-v0.7.6+commit.7338
295f (11260472 bytes)                                                                                                                       
2022-04-14 13:33:57,596 lib.installation INFO     100% of https://binaries.soliditylang.org/linux-amd64/solc-linux-amd64-v0.7.6+commit.73382
95f                                                                                                                                         
2022-04-14 13:33:57,728 lib.installation INFO     Moving from staging (/opt/compiler-explorer/staging/solc-0.7.6) to final destination (/opt
/compiler-explorer/solc-0.7.6)                                                                                                              
2022-04-14 13:33:57,923 lib.installation INFO     compilers/solidity 0.7.6 installed OK                                                     
Installing compilers/solidity 0.8.13                                                                                                        
2022-04-14 13:33:57,966 lib.installation INFO     Fetching https://binaries.soliditylang.org/linux-amd64/solc-linux-amd64-v0.8.13+commit.aba
a5c0e (14156344 bytes)                                                                                                                      
2022-04-14 13:33:58,024 lib.installation INFO     100% of https://binaries.soliditylang.org/linux-amd64/solc-linux-amd64-v0.8.13+commit.abaa
5c0e                                                                                                                                        
2022-04-14 13:33:58,185 lib.installation INFO     Moving from staging (/opt/compiler-explorer/staging/solc-0.8.13) to final destination (/op
t/compiler-explorer/solc-0.8.13)                                                                                                            
2022-04-14 13:33:58,421 lib.installation INFO     compilers/solidity 0.8.13 installed OK                                                    
5 packages installed OK, 0 skipped, and 0 failed installation                                                                               
```
yay

**Comment by anders-torbjornsen** [comment]
@mattgodbolt would solcjs suffice? https://www.npmjs.com/package/solc/v/0.3.6

**Comment by anders-torbjornsen** [comment]
Failing that the source is here at least, so we could probably build it if needs be https://github.com/ethereum/solidity/releases?page=9

**Comment by mattgodbolt** [comment]
> @mattgodbolt would solcjs suffice? https://www.npmjs.com/package/solc/v/0.3.6

I don't know? Is there a single executable `solc` I can pull out of that and install in `/opt/compiler-explorer/solc...` etc? I'd really rather not have to do heroics to do so. @shekhirin any ideas?

**Comment by anders-torbjornsen** [comment]
Nah it's a node package so it would need an nodeJS runtime to execute in etc. I'm just having a go at compiling 0.3.6 from source, hopefully it won't be too bad

**Comment by anders-torbjornsen** [comment]
ok @mattgodbolt I managed to get 0.3.6 compiled on windows, will that do or would you prefer linux?

**Comment by mattgodbolt** [comment]
I don't so much prefer linux as require a linux tarball that runs on an Ubuntu 20.04 machine :

**Comment by mattgodbolt** [comment]
I notice the example code has:
```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity >=0.4.0;
```
too - so if we were to support 0.3.6 we'd presumably have to change that too?

**Comment by anders-torbjornsen** [comment]
Hehe sorry I'm a noob with contributing to this project! I'll fire up wsl tonight and see if I can sort that out.

Good catch on the pragma as well, yeah that'll need changing.

Any plans to support matching blocks of code with asm and highlighting them funky colours etc? I have some experience mapping evm asm back to sol code for other tools in the past so can help out there if needed.

**Comment by shekhirin** [comment]
> I notice the example code has:
> 
> ```solidity
> // SPDX-License-Identifier: UNLICENSED
> pragma solidity >=0.4.0;
> ```
> 
> too - so if we were to support 0.3.6 we'd presumably have to change that too?

pragmas were added in 0.4.0 only https://github.com/ethereum/solidity/blob/develop/Changelog.md#040-2016-09-08, so we need to have this line removed for earlier versions

**Comment by mattgodbolt** [comment]
Thoughts on https://github.com/compiler-explorer/compiler-explorer/issues/3520 welcomed too (this is why I didn't add `solc` support years back when it would have actually been useful for me in my day job :D )

**Comment by anders-torbjornsen** [comment]
Great, I'll throw some comments in there this evening

**Comment by shekhirin** [comment]
> I don't so much prefer linux as require a linux tarball that runs on an Ubuntu 20.04 machine 😀

I compiled 0.3.6 for linux/amd64 from sources and can publish it somewhere on GitHub.
```bash
➜  ~ docker run -it --rm -v $(pwd)/solc:/solc --platform linux/amd64 alpine /solc --version
solc, the solidity compiler commandline interface
Version: 0.3.6-3fc68da5/Release-Linux/g++
```

**Comment by mattgodbolt** [comment]
That would be great; thanks!

**Comment by shekhirin** [comment]
Here it is: https://raw.githubusercontent.com/shekhirin/solc/linux-amd64-v0.3.6/linux/amd64/solc-v0.3.6. I don't think it'll be merged into upstream since it's an old repo and I'm not an employee of this company to introduce new pre-compiled binary.

```console
➜  ~ docker run -it --rm --platform linux/amd64 alpine/curl sh -c "curl -s https://raw.githubusercontent.com/shekhirin/solc/linux-amd64-v0.3.6/linux/amd64/solc-v0.3.6 > solc; chmod +x solc; ./solc --version"
solc, the solidity compiler commandline interface
Version: 0.3.6-3fc68da5/Release-Linux/g++
```

**Comment by anders-torbjornsen** [comment]
Nice one, apologies for not getting to this, busy weekend!

**Pull request merged**